### PR TITLE
Reactant interpreter overload

### PR DESCRIPTION
**Pull request by wsmoses**
hopefully resolves https://github.com/EnzymeAD/Reactant.jl/issues/189

### CODE REVIEW

File: src/utils.jl [hunk 1]
```diff
@@ -120,13 +120,7 @@ function make_mlir_fn(
...
         end
 
-        # NOTE an `AbstractInterpreter` cannot process methods with more recent world-ages than it
-        # solution is to use a new interpreter, but we reuse the `code_cache` to minimize comptime in Julia <= 1.10
-        @static if !HAS_INTEGRATED_CACHE
-            interp = ReactantInterpreter(; code_cache=REACTANT_CACHE)
-        else
-            interp = ReactantInterpreter()
-        end
+        interp = ReactantInterpreter()
```
Comment (mofeing) [inline src/utils.jl:123]: mmm we must keep this for Julia 1.10, because the code cache first arrived to Julia 1.11

Comment (mofeing) [reply src/utils.jl:123]: if we remove this, tracing on Julia 1.10 is gonna slow down a lot

Comment (wsmoses) [reply src/utils.jl:123]: We still have the cache it’s just in the constructor

**Comment by mofeing** [comment]
mmm okay, i see. so with this PR we would be using the interpreter from Enzyme, right?

wouldn't this be problematic? like what if I have a function which I have overlayed for Reactant but then I used that same function in Enzyme without Reactant, or viceversa.

I feel a opposed to removing out interpreter to keeps things clean and separate, but I want to understand it completely before judging this solution.

**Comment by wsmoses** [comment]
> mmm okay, i see. so with this PR we would be using the interpreter from Enzyme, right?
> 

kind of. I revamped the enzyme interpreter to basically be an abstract interpreter with extensions, which is what we do here by providing the custom handle routine. Thus we can successfully override what we want while keeping the other overrides we still need (eg those for custom rules)

**Comment by mofeing** [comment]
okay, let's talk about it on Wednesday's meeting. not against it but just wanna be sure that we don't lose anything.

**Review by mofeing — APPROVED** [review]
Okay, I see nothing wrong in this PR but I would like to add one note for the future: The problem is not in our own `ReactantIntepreter` but that **`AbstractIntepreter` currently doesn't support dynamic dispatch**.

Having our own `AbstractInterpreter` is the way to go, but while this is not supported and as we mainly (and almost only) use it for overlaying `Enzyme.autodiff` (which is critical for us), this PR solves the issue of missing method overlaying by moving the problem to Enzyme. We sacrifice method overlaying for other purposes but it's not something we have exploited a lot.

@wsmoses When method overlaying with dynamic dispatch is fixed, I think we should come back to having our `ReactantInterpreter` for better customization (but this can take a while).

**Comment by wsmoses** [comment]
so partially yes, partially no. Longer term I think we will need something like this to make sure we correctly handle custom derivative rules [or at least have some sharing of the "rule override exists"]

**Comment by mofeing** [comment]
> ... correctly handle custom derivative rules [...]

i think this could be handled by Reactant somehow if `AbstractInterpreter`s compose. anyway, this is kind of speculative and in the end, we'll have to adapt to whatever new changes get added to `AbstractInterpreter`.

mmm it looks like this PR breaks stuff and the "Basic grad cos" test is crashing.

**Comment by wsmoses** [comment]
Yeah this Pr needs a release of enzyme and then rebase atop to run, which I’ll do today now that the plan is approved

**Pull request merged**